File: Tools/Sources/TatamiToolsKit/Localization.swift
// SPDX-FileCopyrightText: 2026 PangMo5 and contributors
// SPDX-License-Identifier: AGPL-3.0-only

import Foundation
import SwiftSoup

// MARK: - TextCatalog

final class TextCatalog {

  // MARK: Lifecycle

  init(values: JSON = .object([]), locale: String = "en") {
    self.values = values
    self.locale = locale
  }

  // MARK: Internal

  let values: JSON
  let locale: String
  var observed = [String: (english: String, context: String)]()

  func text(_ source: String, _ context: String) throws -> String {
    let normalized = trim(replacing(#"\s+"#, in: source) { _ in " " })
    let remainder = trim(replacing(#"\{\d+\}"#, in: normalized) { _ in "" })
    let identity: Set = [
      "Tatami",
      "GitHub",
      "CLI",
      "TOML",
      "BSP",
      "JSON",
      "PangMo5",
      "SwiftyCrow",
      "Amado",
      "macOS",
      "UUID",
      "UUID[]",
      "UUID[]?",
      "double?",
      "table?",
      "string[]?",
      "bool",
      "string",
      "string[]",
      "string?",
      "int",
      "double",
      "table",
      "table[]",
      "en",
      "ko",
      "ja",
      "zh-Hans",
      "zh-Hant",
      "true",
      "false",
      "null",
      "Bool",
      "String",
      "Int",
      "Float",
    ]
    if
      normalized
        .isEmpty || fullMatch(#"\[!(NOTE|TIP|IMPORTANT|WARNING|CAUTION)\]"#, normalized) || matches("[A-Za-z]", remainder)
        .isEmpty || identity.contains(remainder) { return source }
    let key = textKey(normalized)
    if observed[key] == nil { observed[key] = (normalized, context) }
    if locale == "en" { return normalized }
    let translated = values[key][locale].str
    try require(!translated.isEmpty, "Missing \(locale): \(key) / \(normalized)")
    try require(
      matches(#"\{\d+\}"#, normalized).map { $0[0] }.sorted() == matches(#"\{\d+\}"#, translated).map { $0[0] }.sorted(),
      "Changed placeholders: \(key) / \(locale)",
    )
    return translated
  }

}

func htmlEscape(_ source: String, quotes: Bool = true) -> String {
  var text = source.replacingOccurrences(of: "&", with: "&amp;").replacingOccurrences(of: "<", with: "&lt;").replacingOccurrences(
    of: ">",
    with: "&gt;",
  )
  if quotes { text = text.replacingOccurrences(of: "\"", with: "&quot;").replacingOccurrences(of: "'", with: "&#x27;") }
  return text
}

let voidTags: Set = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr",
]

// MARK: - HTMLNode

final class HTMLNode {

  // MARK: Lifecycle

  init(_ tag: String, _ attributes: [(String, String?)] = [], _ children: [HTMLChild] = []) {
    self.tag = tag
    self.attributes = attributes
    self.children = children
  }

  // MARK: Internal

  let tag: String
  var attributes: [(String, String?)]
  var children: [HTMLChild]

  subscript(_ attribute: String) -> String? {
    get { attributes.first { $0.0 == attribute }?.1 }
    set {
      if let index = attributes.firstIndex(where: { $0.0 == attribute }) { attributes[index].1 = newValue }
      else { attributes.append((attribute, newValue)) }
    }
  }

  func has(_ attribute: String) -> Bool {
    attributes.contains { $0.0 == attribute }
  }

  func render() -> String {
    let content = children.map { $0.render() }.joined()
    switch tag {
    case "#root": return content
    case "#markup": return children.first?.text ?? ""
    case "#comment": return "<!--" + (children.first?.text ?? "") + "-->"
    case "#doctype": return "<!" + (children.first?.text ?? "") + ">"
    default: break
    }
    let attrs = attributes.map { " " + $0.0 + ($0.1.map { "=\"" + htmlEscape($0) + "\"" } ?? "") }.joined()
    let opening = "<" + tag + attrs + ">"
    if voidTags.contains(tag) { return opening }
    return opening + (["script", "style"].contains(tag) ? children.compactMap(\.text).joined() : content) + "</" + tag + ">"
  }

  func walk(_ visitor: (HTMLNode) throws -> Void) rethrows {
    try visitor(self)
    for child in children { if let node = child.node { try node.walk(visitor) } }
  }

}

// MARK: - HTMLChild

enum HTMLChild {
  case text(String)
  case node(HTMLNode)

  var text: String? {
    if case .text(let value) = self { return value }
    return nil
  }

  var node: HTMLNode? {
    if case .node(let value) = self { return value }
    return nil
  }

  func render() -> String {
    switch self
    { case .text(let value): htmlEscape(value, quotes: false)
    case .node(let value): value.render() }
  }
}

/// SwiftSoup handles HTML tokenization, entities and the browser-compatible tree.
/// The small translation tree keeps movable prose placeholders separate from markup.
func translationTree(_ source: String) throws -> HTMLNode {
  func convert(_ node: SwiftSoup.Node) throws -> HTMLChild {
    if let node = node as? SwiftSoup.TextNode { return .text(node.getWholeText()) }
    if let node = node as? SwiftSoup.DataNode { return .text(node.getWholeData()) }
    if let node = node as? SwiftSoup.Comment { return .node(HTMLNode("#comment", [], [.text(node.getData())])) }
    if node is SwiftSoup.DocumentType { return .node(HTMLNode("#markup", [], [.text(try node.outerHtml())])) }
    let booleanAttributes: Set = [
      "controls",
      "muted",
      "playsinline",
      "defer",
      "selected",
      "disabled",
      "hidden",
      "autoplay",
      "loop",
      "checked",
      "multiple",
      "required",
      "open",
    ]
    let attributes = node.getAttributes()?.asList().map { attribute -> (String, String?) in
      let key = attribute.getKey()
      let value = attribute.getValue()
      return (key, booleanAttributes.contains(key) && value.isEmpty ? nil : value)
    } ?? []
    return .node(HTMLNode(node.nodeName(), attributes, try node.getChildNodes().map(convert)))
  }
  let nodes: [SwiftSoup.Node] =
    if !matches("(?i)<(?:!doctype|html)\\b", source).isEmpty {
      try SwiftSoup.parse(source).getChildNodes()
    } else {
      try SwiftSoup.parseBodyFragment(source).body()?.getChildNodes() ?? []
    }
  return HTMLNode("#root", [], try nodes.map(convert))
}

func translateHTML(_ source: String, _ catalog: TextCatalog, _ context: String) throws -> HTMLNode {
  let root = try translationTree(source)
  let blocks: Set = ["div", "section", "nav", "aside", "article", "main", "footer", "header", "ul", "ol", "li", "video", "pre"]
  let translatable: Set = [
    "p",
    "h1",
    "h2",
    "h3",
    "h4",
    "title",
    "a",
    "button",
    "summary",
    "label",
    "span",
    "option",
    "sub",
    "small",
  ]
  func visit(_ node: HTMLNode) throws {
    if ["script", "style", "pre", "#comment", "#doctype"].contains(node.tag) { return }
    for attribute in [
      "aria-label",
      "title",
      "alt",
      "placeholder",
      "data-play-hint",
      "data-copied-label",
      "data-copy-error",
      "data-load-error",
      "data-no-notes",
    ] {
      if let value = node[attribute], !value.isEmpty { node[attribute] = try catalog.text(value, context + " / " + attribute) }
    }
    if node.tag == "meta" && node["name"] == "description", let content = node["content"] { node["content"] = try catalog.text(
      content,
      context + " / meta",
    ) }
    for child in node.children { if let child = child.node { try visit(child) } }
    let eligible = translatable.contains(node.tag) || (node.tag == "div" && node.children.allSatisfy { $0.text != nil })
    if !eligible || node.children.contains(where: { $0.node.map { blocks.contains($0.tag) } ?? false }) { return }
    var protected = [HTMLChild]()
    var source = ""
    func protect(_ child: HTMLChild) {
      source += "{\(protected.count)}"
      protected.append(child)
    }
    for child in node.children {
      if
        let element = child.node, ["em", "strong", "b", "i"].contains(element.tag),
        element.children.allSatisfy({ $0.text != nil })
      {
        protect(.node(HTMLNode("#markup", [], [.text("<" + element.tag + ">")])))
        source += element.children.compactMap(\.text).joined()
        protect(.node(HTMLNode("#markup", [], [.text("</" + element.tag + ">")])))
      } else if child.node != nil { protect(child) }
      else { source += child.text! }
    }
    let translated = try catalog.text(source, context + " / " + node.tag)
    var children = [HTMLChild]()
    var cursor = translated.startIndex
    let regex = try! NSRegularExpression(pattern: #"\{(\d+)\}"#)
    for match in regex.matches(in: translated, range: NSRange(translated.startIndex..., in: translated)) {
      let range = Range(match.range, in: translated)!
      if cursor < range.lowerBound { children.append(.text(String(translated[cursor..<range.lowerBound]))) }
      let index = Int(translated[Range(match.range(at: 1), in: translated)!])!
      try require(index < protected.count, "Invalid protected HTML index")
      children.append(protected[index])
      cursor = range.upperBound
    }
    if cursor < translated.endIndex { children.append(.text(String(translated[cursor...]))) }
    node.children = children
  }
  try visit(root)
  return root
}

func inlineUnit(_ source: String, _ catalog: TextCatalog, _ context: String) throws -> String {
  var protected = [String]()
  let text = replacing(#"<[^>]+>|(`+)(.+?)\1|(?<=\]\()([^\s)]+)(?=\))|https?://[^\s<>]+"#, in: source) { match in
    protected.append(match[0])
    return "{\(protected.count - 1)}"
  }
  return try replacing(#"\{(\d+)\}"#, in: catalog.text(text, context)) { match in
    let index = Int(match[1])!
    try require(index < protected.count, "Invalid inline placeholder")
    return protected[index]
  }
}

func markdownUnits(_ source: String, _ catalog: TextCatalog, _ context: String) throws -> String {
  let input = lines(source)
  var result = [String]()
  var index = 0
  var fence: Character?
  var referenceMatrix = false
  while index < input.count {
    let line = input[index]
    if trim(line).isEmpty { referenceMatrix = false }
    if context.hasSuffix("LOCALIZATION.md") && line.hasPrefix("| Concept |") { referenceMatrix = true }
    if let marker = matches(#"^\s*(`{3,}|~{3,})"#, line).first {
      if fence == nil { fence = marker[1].first } else if fence == marker[1].first { fence = nil }
      result.append(line)
      index += 1
      continue
    }
    if
      fence != nil || trim(line).isEmpty || line.hasPrefix("<!--") || !matches(#"^\[[^]]+\]:"#, line).isEmpty || fullMatch(
        #"\s*</?details(?:\s[^>]*)?>\s*"#,
        line,
      )
    {
      result.append(line)
      index += 1
      continue
    }
    if trim(line).hasPrefix("<") {
      var block = [line]
      index += 1
      if !matches(#"^<(p|div)\b"#, trim(line)).isEmpty {
        while index < input.count, matches(#"</(?:p|div)>"#, block.last!).isEmpty { block.append(input[index])
          index += 1
        }
      }
      result.append(try translateHTML(block.joined(separator: "\n"), catalog, context + " / HTML").render())
      continue
    }
    if line.hasPrefix("> ") {
      let text = String(line.dropFirst(2))
      index += 1
      if fullMatch(#"\[![A-Z]+\]"#, text) { result.append(line)
        continue
      }
      var pieces = [text]
      while
        index < input.count, input[index].hasPrefix("> "),
        !trim(String(input[index].dropFirst(2))).isEmpty
      { pieces.append(String(input[index].dropFirst(2)))
        index += 1
      }
      result.append(try "> " + inlineUnit(pieces.joined(separator: " "), catalog, context + " / quote"))
      continue
    }
    if let heading = matches(#"^(#{1,6})\s+(.+)$"#, line).first {
      result.append(try heading[1] + " " + inlineUnit(heading[2], catalog, context + " / heading"))
      index += 1
      continue
    }
    if line.hasPrefix("|") {
      if fullMatch(#"[| :\-]+"#, line) { result.append(line) }
      else {
        let separated = replacing(#"\|(?=(?:[^`]*`[^`]*`)*[^`]*$)"#, in: line) { _ in "\u{0}" }.components(separatedBy: "\u{0}")
        let cells = try separated.enumerated().map { position, raw in
          let text = trim(raw)
          if referenceMatrix && position >= 2 { return text }
          if context == "DemoLab/README.md" {
            if position == 1, ["Design", "Write", "Review", "Chat", "Build", "Focus"].contains(text) { return text }
            if
              position == 2, text.components(separatedBy: "+").allSatisfy({ [
                "Canvas",
                "Docs",
                "Editor",
                "Review",
                "Chat",
                "Terminal",
                "Notes",
                "Monitor",
              ].contains(trim($0)) }) { return text }
          }
          return try text.isEmpty ? "" : inlineUnit(text, catalog, context + " / table")
        }
        result.append(cells.joined(separator: "|"))
      }
      index += 1
      continue
    }
    let prefix = matches(#"^(\s*(?:[-*+] |\d+\. )|> )(.*)$"#, line).first
    let lead = prefix?[1] ?? ""
    var block = [prefix?[2] ?? line]
    index += 1
    while
      index < input.count, !trim(input[index]).isEmpty, matches(
        #"^(?:#{1,6} |\s*[-*+] |\s*\d+\. |\||<|>|```|~~~)"#,
        input[index],
      ).isEmpty
    { block.append(trim(input[index]))
      index += 1
    }
    result.append(try lead + inlineUnit(block.joined(separator: " "), catalog, context + " / prose"))
  }
  return result.joined(separator: "\n") + "\n"
}

File: scripts/localization.py
# SPDX-FileCopyrightText: 2026 PangMo5 and contributors
# SPDX-License-Identifier: AGPL-3.0-only
"""Build-time text units. Code, links and markup remain structural data."""
from dataclasses import dataclass, field
from html.parser import HTMLParser
from pathlib import Path
import hashlib
import html
import json
import re

ROOT = Path(__file__).resolve().parents[1]
LOCALES = ['en', 'ko', 'ja', 'zh-Hans', 'zh-Hant']
NAMES = dict(zip(LOCALES, ['English', '한국어', '日本語', '简体中文', '繁體中文']))
IDENTITY = {'Tatami','GitHub','CLI','TOML','BSP','JSON','PangMo5','SwiftyCrow','Amado','macOS',
            'UUID','UUID[]','UUID[]?','double?','table?','string[]?','bool','string','string[]','string?','int','double','table','table[]',
            'en','ko','ja','zh-Hans','zh-Hant','true','false','null','Bool','String','Int','Float'}


def key(text):
    return hashlib.sha256(text.encode()).hexdigest()[:16]


def human(text):
    if re.fullmatch(r'\[!(NOTE|TIP|IMPORTANT|WARNING|CAUTION)\]', text): return False
    remainder = re.sub(r'\{\d+\}', '', text).strip()
    return bool(re.search('[A-Za-z]', remainder)) and remainder not in IDENTITY


@dataclass
class TextCatalog:
    values: dict = field(default_factory=dict)
    observed: dict = field(default_factory=dict)
    locale: str = 'en'

    def text(self, text, context):
        normalized = re.sub(r'\s+', ' ', text).strip()
        if not normalized or not human(normalized): return text
        identifier = key(normalized)
        self.observed.setdefault(identifier, {'en': normalized, 'context': context})
        if self.locale == 'en': return normalized
        value = self.values.get(identifier, {}).get(self.locale)
        if not value: raise ValueError(f'Missing {self.locale}: {identifier} / {normalized}')
        if sorted(re.findall(r'\{\d+\}', normalized)) != sorted(re.findall(r'\{\d+\}', value)):
            raise ValueError(f'Changed placeholders: {identifier} / {self.locale}')
        return value


VOID = {'area','base','br','col','embed','hr','img','input','link','meta','param','source','track','wbr'}
BLOCK = {'div','section','nav','aside','article','main','footer','header','ul','ol','li','video','pre'}
TRANSLATABLE = {'p','h1','h2','h3','h4','title','a','button','summary','label','span','option','sub','small'}

@dataclass
class Node:
    tag: str
    attrs: dict = field(default_factory=dict)
    children: list = field(default_factory=list)

    def render(self):
        if self.tag == '#root': return ''.join(render(c) for c in self.children)
        if self.tag == '#markup': return self.children[0]
        if self.tag == '#comment': return '<!--'+self.children[0]+'-->'
        if self.tag == '#doctype': return '<!'+self.children[0]+'>'
        attrs = ''.join(' '+k+(f'="{html.escape(v,quote=True)}"' if v is not None else '') for k,v in self.attrs.items())
        opening = '<'+self.tag+attrs+'>'
        if self.tag in VOID: return opening
        content = ''.join(self.children) if self.tag in {'script','style'} else ''.join(render(c) for c in self.children)
        return opening+content+'</'+self.tag+'>'


def render(value): return value.render() if isinstance(value,Node) else html.escape(value,quote=False)

class Document(HTMLParser):
    def __init__(self, source):
        super().__init__(convert_charrefs=True)
        self.root=Node('#root');self.stack=[self.root];self.feed(source)
    def handle_starttag(self, tag, attrs):
        node=Node(tag,dict(attrs));self.stack[-1].children.append(node)
        if tag not in VOID:self.stack.append(node)
    def handle_startendtag(self, tag, attrs): self.handle_starttag(tag,attrs)
    def handle_endtag(self, tag):
        if len(self.stack)>1 and self.stack[-1].tag==tag:self.stack.pop()
        elif tag not in VOID:raise ValueError(f'Mismatched closing tag: {tag}')
    def handle_data(self, data): self.stack[-1].children.append(data)
    def handle_comment(self, data):self.stack[-1].children.append(Node('#comment',children=[data]))
    def handle_decl(self, decl):self.stack[-1].children.append(Node('#doctype',children=[decl]))


def translate_html(source, catalog, context):
    document=Document(source)
    def visit(node):
        if node.tag in {'script','style','pre','#comment','#doctype'}:return
        for attribute in ['aria-label','title','alt','placeholder','data-play-hint','data-copied-label','data-copy-error','data-load-error','data-no-notes']:
            if node.attrs.get(attribute):node.attrs[attribute]=catalog.text(node.attrs[attribute],context+' / '+attribute)
        if node.tag=='meta' and node.attrs.get('name')=='description':node.attrs['content']=catalog.text(node.attrs['content'],context+' / meta')
        for child in node.children:
            if isinstance(child,Node):visit(child)
        eligible = node.tag in TRANSLATABLE or (node.tag=='div' and all(isinstance(c,str) for c in node.children))
        if not eligible or any(isinstance(c,Node) and c.tag in BLOCK for c in node.children):return
        protected=[];text=''
        for child in node.children:
            if isinstance(child,Node) and child.tag in {'em','strong','b','i'} and all(isinstance(c,str) for c in child.children):
                text+='{'+str(len(protected))+'}'+''.join(child.children);protected.append(Node('#markup',children=['<'+child.tag+'>']))
                text+='{'+str(len(protected))+'}';protected.append(Node('#markup',children=['</'+child.tag+'>']))
            elif isinstance(child,Node):
                text+='{'+str(len(protected))+'}';protected.append(child)
            else:text+=child
        localized=catalog.text(text,context+' / '+node.tag)
        children=[]
        for token in re.split(r'(\{\d+\})',localized):
            if re.fullmatch(r'\{\d+\}',token):children.append(protected[int(token[1:-1])])
            elif token:children.append(token)
        node.children=children
    visit(document.root)
    return document.root


# Inline source code and link destinations are opaque to the prose catalog.
INLINE=re.compile(r'<[^>]+>|(`+)(.+?)\1|(?<=\]\()([^\s)]+)(?=\))|https?://[^\s<>]+')

def inline_unit(text, catalog, context):
    protected=[]
    def protect(match):
        protected.append(match.group(0));return '{'+str(len(protected)-1)+'}'
    source=INLINE.sub(protect,text)
    localized=catalog.text(source,context)
    return re.sub(r'\{(\d+)\}',lambda m:protected[int(m.group(1))],localized)


def markdown_units(source, catalog, context):
    lines=source.splitlines();result=[];index=0;fence=None;reference_matrix=False
    while index<len(lines):
        line=lines[index]
        if not line.strip(): reference_matrix=False
        if context.endswith('LOCALIZATION.md') and line.startswith('| Concept |'): reference_matrix=True
        marker=re.match(r'^\s*(`{3,}|~{3,})',line)
        if marker:
            if fence is None:fence=marker.group(1)[0]
            elif marker.group(1)[0]==fence:fence=None
            result.append(line);index+=1;continue
        if fence or not line.strip() or line.startswith('<!--') or re.match(r'^\[[^]]+\]:',line):
            result.append(line);index+=1;continue
        if re.fullmatch(r'\s*</?details(?:\s[^>]*)?>\s*', line):
            result.append(line);index+=1;continue
        if line.lstrip().startswith('<'):
            # HTML blocks (README screenshots/disclosures) use the same DOM rules.
            block=[line];index+=1
            if re.match(r'^<(p|div)\b',line.strip()):
                while index<len(lines) and not re.search(r'</(?:p|div)>',block[-1]):block.append(lines[index]);index+=1
            try:result.append(translate_html('\n'.join(block),catalog,context+' / HTML').render())
            except ValueError:
                # Opening/closing disclosure tags carry no prose and span MD blocks.
                if all(re.fullmatch(r'\s*</?details>\s*',x) for x in block):result.extend(block)
                else:raise
            continue
        if line.startswith('> '):
            text=line[2:];index+=1
            if re.fullmatch(r'\[![A-Z]+\]',text):result.append(line);continue
            pieces=[text]
            while index<len(lines) and lines[index].startswith('> ') and lines[index][2:].strip():
                pieces.append(lines[index][2:]);index+=1
            result.append('> '+inline_unit(' '.join(pieces),catalog,context+' / quote'));continue
        heading=re.match(r'^(#{1,6})\s+(.+)$',line)
        if heading:
            result.append(heading.group(1)+' '+inline_unit(heading.group(2),catalog,context+' / heading'));index+=1;continue
        if line.startswith('|'):
            if re.fullmatch(r'[| :\-]+',line):result.append(line)
            else:
                # Pipes inside inline code belong to the sample, not the table.
                cells=re.split(r'\|(?=(?:[^`]*`[^`]*`)*[^`]*$)',line)
                def table_cell(i, cell):
                    text=cell.strip()
                    if reference_matrix and i>=2: return text
                    if context=='DemoLab/README.md':
                        if i==1 and text in {'Design','Write','Review','Chat','Build','Focus'}: return text
                        if i==2 and all(x.strip() in {'Canvas','Docs','Editor','Review','Chat','Terminal','Notes','Monitor'} for x in text.split('+')): return text
                    return inline_unit(text,catalog,context+' / table') if text else ''
                result.append('|'.join(table_cell(i,c) for i,c in enumerate(cells)))
            index+=1;continue
        prefix=re.match(r'^(\s*(?:[-*+] |\d+\. )|> )(.*)$',line)
        lead=prefix.group(1) if prefix else '';block=[prefix.group(2) if prefix else line];index+=1
        while index<len(lines) and lines[index].strip() and not re.match(r'^(?:#{1,6} |\s*[-*+] |\s*\d+\. |\||<|>|```|~~~)',lines[index]):
            block.append(lines[index].strip());index+=1
        result.append(lead+inline_unit(' '.join(block),catalog,context+' / prose'))
    return '\n'.join(result)+'\n'
